### PR TITLE
Child logout

### PR DESCRIPTION
**Pull request by AryamanDubey28**
Hi Guys,

This is the implementation of the child mode log out task. Log out button in child mode is clicked and then they have to enter the PIN to actually log out back to the admin home

Let me know if there's any issues!

Thanks,
Aryaman

### CODE REVIEW

File: lib/pages/admin/admin_interface.dart [hunk 1]
```diff
@@ -12,6 +12,12 @@ class AdminInterface extends StatelessWidget {
       theme: ThemeData(
         primarySwatch: Colors.teal,
       ),
+      // routes: {
```
Comment (skellyhoe) [inline lib/pages/admin/admin_interface.dart]: delete commented out code

File: lib/pages/child_board/child_board_interface.dart [hunk 2]
```diff
@@ -14,6 +15,13 @@ class ChildInterface extends StatelessWidget {
       theme: ThemeData(
         primarySwatch: Colors.teal,
       ),
+      routes: {
+        // When navigating to the "homeScreen" route, build the HomeScreen widget.
+        'homeScreen': (context) => ChildBoards(),
+        'adminScreen': (context) => AdminChoiceBoards(),
```
Comment (skellyhoe) [inline lib/pages/child_board/child_board_interface.dart]: this whole file will be deleted make sure it has nothing important

File: lib/pages/child_menu/customizable_column.dart [hunk 3]
```diff
@@ -37,10 +58,72 @@ class CustomizableColumn extends StatelessWidget {
...
+  void submit(BuildContext context) {
+    //verifys password is correct, if so then navigates back. otherwise says incorrect
+    if (pin_controller.text.trim() == "0000") {
+      print("navigating back to admin home");
+      Navigator.push(
+        context,
+        MaterialPageRoute(builder: (context) => AdminChoiceBoards()),
+      );
+
+      //Navigator.popAndPushNamed(context, 'adminScreen');
```
Comment (skellyhoe) [inline lib/pages/child_menu/customizable_column.dart]: remove commented out code

File: lib/pages/child_menu/customizable_column.dart [hunk 4]
```diff
@@ -37,10 +58,72 @@ class CustomizableColumn extends StatelessWidget {
...
+                  onPressed: () => submit(context),
+                  child: Text("SUBMIT"))
+            ],
+          ));
+
+  void submit(BuildContext context) {
+    //verifys password is correct, if so then navigates back. otherwise says incorrect
+    if (pin_controller.text.trim() == "0000") {
+      print("navigating back to admin home");
+      Navigator.push(
```
Comment (skellyhoe) [inline lib/pages/child_menu/customizable_column.dart]: use pushAndReplace instead of push

File: lib/pages/child_menu/customizable_column.dart [hunk 5]
```diff
@@ -37,10 +58,72 @@ class CustomizableColumn extends StatelessWidget {
...
+        title: Text("Child Mode"),
+        automaticallyImplyLeading: false,
+        actions: [
+          Padding(
+            padding: const EdgeInsets.only(right: 20.0),
+            child: GestureDetector(
+                key: Key("logoutButton"),
+                //only triggers when its pressed for some time and swiped up
+                // onLongPressUp: () {
+                //   openLogoutDialog(context);
```
Comment (skellyhoe) [inline lib/pages/child_menu/customizable_column.dart]: commented out code

File: lib/widgets/logout_alert_dialog.dart [hunk 6]
```diff
@@ -0,0 +1,12 @@
+import 'package:flutter/material.dart';
+import 'package:flutter/src/widgets/framework.dart';
+import 'package:flutter/src/widgets/placeholder.dart';
+
```
Comment (skellyhoe) [inline lib/widgets/logout_alert_dialog.dart]: what does this widget do?

File: lib/pages/child_menu/customizable_column.dart [hunk 7]
```diff
@@ -37,10 +58,72 @@ class CustomizableColumn extends StatelessWidget {
...
+                  key: Key("submitButton"),
+                  onPressed: () => submit(context),
+                  child: Text("SUBMIT"))
+            ],
+          ));
+
+  void submit(BuildContext context) {
+    //verifys password is correct, if so then navigates back. otherwise says incorrect
+    if (pin_controller.text.trim() == "0000") {
+      print("navigating back to admin home");
```
Comment (skellyhoe) [inline lib/pages/child_menu/customizable_column.dart]: remove print statement

**Review by skellyhoe — CHANGES_REQUESTED** [review]
Good work!

Some notes:

- Remove commented out code highlighted.
- Write general comments on methods and pages you created.
- Remove print statement highlighted.
- Use pushAndReplace instead of push where highlighted.
- You wrote code in child_board_interface, but that file will be deleted. Make sure if the code is needed to add it somewhere else.

For the log out button consider these changes:
- When the admin presses the activate child view for the first time it should ask the admin for a PIN and then it saves it and doesn't ask for it again.
- It's still set to a tap gesture and not a hold and swipe gesture.
- Instead of a whole keyboard, the user should be presented with a number pad. 
- Instead of a whole text box, the user should be presented with a simple dialog as explained on WhatsApp.

File: lib/main.dart [hunk 8]
```diff
@@ -1,21 +1,32 @@
...
 
   @override
   Widget build(BuildContext context) {
     return MaterialApp(
       debugShowCheckedModeBanner: false,
-      home: Wrapper(),
+      theme: ThemeData(
+        primarySwatch: Colors.teal,
+      ),
+      //home: Wrapper(),
```
Comment (Ali-Alkhars) [inline lib/main.dart]: Remove commented out code

File: lib/pages/admin/admin_side_menu.dart [hunk 9]
```diff
@@ -37,25 +38,36 @@ class AdminSideMenu extends StatelessWidget {
...
               )),
             ),
             ListTile(
               key: const Key("childMode"),
               leading: const Icon(Icons.child_care),
               title: const Text('Activate Child Mode'),
-              onTap: () => Navigator.of(context).pushReplacement(MaterialPageRoute(
-                builder: (context) => CustomizableColumn(),
-              )),
+              // onTap: () => Navigator.of(context).pushReplacement(MaterialPageRoute(
```
Comment (Ali-Alkhars) [inline lib/pages/admin/admin_side_menu.dart]: Remove commented out code

File: lib/pages/authenticate/wrapper.dart [hunk 10]
```diff
@@ -16,7 +19,8 @@ class Wrapper extends StatelessWidget {
         builder: (context, snapshot) {
           if (snapshot.hasData) {
             //if we are trying to sign in and snapshot contains user data, we are logged in
-            return const AdminInterface();
+            return isInChildMode ? CustomizableColumn() : AdminInterface();
+            //return const AdminInterface();
```
Comment (Ali-Alkhars) [inline lib/pages/authenticate/wrapper.dart]: Remove commented out code

File: lib/pages/child_menu/customizable_column.dart [hunk 11]
```diff
@@ -37,10 +60,74 @@ class CustomizableColumn extends StatelessWidget {
...
+              TextButton(
+                  key: Key("submitButton"),
+                  onPressed: () => submit(context),
+                  child: Text("SUBMIT"))
+            ],
+          ));
+
+  Future<void> submit(BuildContext context) async {
+    //verifys password is correct, if so then navigates back. otherwise says incorrect
+    if (pin_controller.text.trim() == "0000") {
```
Comment (Ali-Alkhars) [inline lib/pages/child_menu/customizable_column.dart]: Why is the password hardcoded?

File: lib/widgets/clickable_images_grid.dart [hunk 12]
```diff
@@ -17,14 +18,33 @@ Expanded getMainImages(List<ClickableImage> images) {
   );
 }
 
+Future speak(String text) async {
```
Comment (Ali-Alkhars) [inline lib/widgets/clickable_images_grid.dart:24]: Add a doc comment explaining what the method does

**Review by Ali-Alkhars — CHANGES_REQUESTED** [review]
Hello, nice work overall!

I have some minor comments you could see below. 

**Other than that I have one comment,** I know that integration tests are cool and everything, but they're slow and kinda shi**y ngl haha. I don't really see a reason why you can't do the same tests as widget tests

**Comment by AryamanDubey28** [comment]
Hi guys,
Apologies for the delays but I've incorporated all the changes so check them out when you can!

File: lib/widgets/edit_pin_section.dart [hunk 13]
```diff
@@ -0,0 +1,166 @@
...
+                  onPressed: () => submit(context),
+                  child: Text("SUBMIT"))
+            ],
+          ));
+
+  Future<void> submit(BuildContext context) async {
+    String result =
+        await authentitcationHelper.createPIN(_pinEditController.text.trim());
+    show_alert_dialog(result);
+    //Navigator.of(context).pop();
```
Comment (skellyhoe) [inline lib/widgets/edit_pin_section.dart]: commented out code

File: lib/widgets/edit_pin_section.dart [hunk 14]
```diff
@@ -0,0 +1,166 @@
...
+/// This widget returns all the components and functionalitlies necessary for the user to change their email.
+class EditPINSection extends StatelessWidget {
+  late BuildContext context;
+  final _pinEditController = TextEditingController();
+  late final Auth authentitcationHelper;
+
+  EditPINSection({required this.authentitcationHelper});
+
+  // Displays an alert dialog with the text passed as parameter.
+  void show_alert_dialog(String text) {
```
Comment (skellyhoe) [inline lib/widgets/edit_pin_section.dart:18]: Duplicate method. Make it a separate widget if possible.

File: lib/main.dart [hunk 15]
```diff
@@ -27,8 +35,12 @@ class MyApp extends StatelessWidget {
 
     return MaterialApp(
       debugShowCheckedModeBanner: false,
-      theme: themeData,
-      home: Wrapper(),
+      theme: ThemeData(
+        primarySwatch: Colors.teal,
+      ),
+      //home: Wrapper(),
```
Comment (skellyhoe) [inline lib/main.dart]: still commented out code here

File: lib/pages/admin/admin_side_menu.dart [hunk 16]
```diff
@@ -38,7 +41,22 @@ class AdminSideMenu extends StatelessWidget {
         ),
       );
 
+  // Displays an alert dialog with the text passed as parameter.
+  void show_alert_dialog(BuildContext context, String text) {
```
Comment (skellyhoe) [inline lib/pages/admin/admin_side_menu.dart:46]: Duplicate method. Make it a separate widget if possible.

**Review by skellyhoe — APPROVED** [review]
Nice job!

I left some comments down. For the dialog widget, if it doesn't work then don't implement it. Make sure to pull from main to get the most up to date changes and see if they conflict with your work. I'll approve it now but make sure to delete the commented out code first. Also check with the others to see if we're still doing integration tests.

**Review by Ali-Alkhars — CHANGES_REQUESTED** [review]
- For some reason the themes don't work on the admin's side
- Users can enter a text pin from the edit account page and it gets accepted! (the keyboard should be only digits & it should show an error message if it was anything other than digits)
- Try clicking on the change pin text. The keyboard covers the text field.
- I wasn't asked to enter a pin. When I clicked on child app without setting a pin I was taken there and got stuck!

**Review by absirgue — CHANGES_REQUESTED** [review]
Hi Aryaman,

Thanks for the work. I like the hold button and the new edit account page.

First (as shown by screenshots attached), firebase errors are passed in to the prompt for the pin input field in the edit account page. This should abolsutely be avoided (show a different UI when my pin is not yet configured, will be way better + actually catch those errors and never display them to the user as raw as this).

Second, the custom theme feature fails on your branch. It works on main and everywhere else, please find the reason why it does not on ur branch (probably a failed merge conflict resolution) and correct it so that the custom themes are exactly working as they do on main (just copy main from main and do some of ur changes again if needed). 

Third, the testing does not cover enough cases (as illustrated by the fact that the other points of the reviews are coming up only in manual testing). Please test:
- pins longer than 4 digits
- pins that don't only contain numbers
- you can not navigate to child app if you didn't set a pin

Fourth, there is a significant UX problem: the admin can go in the child app before having even set a pin. This does not make much sense and should be avoided (show na alert dialog if you don't have set a pin yet and u try to go to the child app asking to first set a pin).

Five, you can put letters in your pin which 1- is not what we want. 2- you can not input letters in the numeric keyboard when trying to go from child app to admin app. Please change the keyboard for pin creation to be the same numeric keyboard

I know this is not the expected review as we were all hoping to be ready to merge this task. It is now 13 days late. Please integrate all of these changes ASAP and do manual testing on your own for a significant time before asking for reviews again. Your sprint 3 task is big and this task now needs to be done in a nice way. 
Screen Shot 2023-03-10 at 6 07 05 PM
Screen Shot 2023-03-10 at 6 07 34 PM

Thanks

**Review by dmunipi1 — APPROVED** [review]
Well done Aryaman

**Pull request merged**